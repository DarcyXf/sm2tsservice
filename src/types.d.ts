import { JSONSchema4, JSONSchema6 } from 'json-schema';
import { IncomingMessage, ServerResponse } from 'http';

export interface PathJson {
  parameters: {
    name: string;
    in: 'path' | 'form' | 'query' | 'body' | string;
    description: string;
    required: boolean;
    type?: string;
    schema?: JSONSchema4 | JSONSchema6;
  }[];
  responses: {
    [status: number]: JSONSchema4 | JSONSchema6;
  };
}

export interface SwaggerJson {
  __mtime?: any;
  paths: {
    [path: string]: {
      [method: string]: PathJson;
    };
  };
  definitions?: JSONSchema6;
  basePath: string;
}

export interface SMAjaxConfig {
  method: 'GET' | 'POST' | 'PUT' | 'DELETE' | 'OPTIONS' | 'PATCH' | 'HEAD';
  url: string;
  data?: any; // post json
  form?: any; // post form
  query?: any;
  header?: any;
  path?: any;
}

export interface SMValidateInfo {
  swagger: { path: string; basePath: string };
  send?: SMAjaxConfig;
  receive: {
    body?: any;
    status: number;
  };
}

export type SMAbstractRequest = IncomingMessage;

export type SMAbstractResponse = ServerResponse | IncomingMessage;

export type SMAbstractNext = (...args: any[]) => any;

export type SMValidator = (info: { code: number; message?: string; result: SMValidateInfo }) => any;
