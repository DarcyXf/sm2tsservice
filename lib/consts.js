'use strict';
Object.defineProperty(exports, '__esModule', { value: true });
var path = require('path');
exports.pluginsPath = path.join(__dirname, '..', 'plugins');
exports.generatorPath = path.join(exports.pluginsPath, 'swagger-codegen-cli.jar');
//# sourceMappingURL=consts.js.map
