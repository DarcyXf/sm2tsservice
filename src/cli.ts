#!/usr/bin/env node
import * as fs from 'fs';
import * as path from 'path';
import * as commander from 'commander';
import gen from './index';

export interface Json2Service {
  url: string;
  type?: 'yapi' | 'swagger';
  yapiConfig?: {
    required?: boolean;
  };
  swaggerParser: SwaggerParser;
}

export interface SwaggerParser {
  '-o'?: string;
  '-t'?: string;
  '-l'?: string;
  '-i': string;
}

const CD = process.cwd();

commander
  .version(require('../package.json').version)
  .option('-c, --config [type]', 'config file', 'json2service.json')
  .option('--clear [type]', 'rm typescript service before gen', false)
  .parse(process.argv);

const Config = commander.config as string;
const ConfigFile = path.join(CD, Config);

if (!fs.existsSync(ConfigFile)) {
  console.error(`[ERROR]: ${Config} not found in ${CD}`);
} else {
  const config: Json2Service = require(ConfigFile);
  gen(config, { clear: commander.clear });
}
