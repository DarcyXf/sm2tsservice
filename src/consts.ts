import * as path from 'path';
import * as fs from 'fs';

export const pluginsPath = path.join(__dirname, '..', 'plugins');
export const generatorPath = path.join(pluginsPath, 'swagger-codegen-cli.jar');

export const DefaultBasePath = '@empty@';
export const SmTmpDir = path.join(__dirname, '..', 'tmp');

if (!fs.existsSync(SmTmpDir)) {
  fs.mkdirSync(SmTmpDir);
}
