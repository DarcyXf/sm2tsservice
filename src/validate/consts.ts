export const X_SM_PATH = 'x-sm-path';
export const X_SM_BASEPATH = 'x-sm-basepath';

export const X_SM_PARAMS = 'x-sm-params';
export const X_SM_ERROR = 'x-sm-error';
