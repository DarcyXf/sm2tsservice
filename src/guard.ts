import { SwaggerJson } from './consts';

/**
 * @description 对 swagger 返回做安全检测
 * + operationId 检测，如果发现 [*]_1 结构，抛出错误，要求前端显示配置
 */

type API = SwaggerJson['paths']['a']['get'];

export const dangerousOperationIdReg = /_[0-9]{1,}$/g;
/**
 *
 * @param swagger json
 * @param savedOperationId2UrlMap  oldOperationId: url 安全映射，用以校正风险 swagger
 */
export function operationIdGuard(
  swagger: SwaggerJson,
  savedOperationId2UrlMap: { [id: string]: string } = {}
) {
  const { paths } = swagger;
  const operationId2UrlMap: { [id: string]: string } = {};
  const operationId2ApiMap: { [id: string]: API } = {};
  const dangerousOperationId2ApiMap: { [id: string]: API[] } = {};
  const url2ApiMap: { [url: string]: API } = {};
  const copySavedOperationId2UrlMap = { ...savedOperationId2UrlMap };
  return new Promise<{
    errors: string[];
    warnings: string[];
    suggestions: typeof operationId2UrlMap;
  }>(rs => {
    const dangers = Object.keys(
      Object.keys(paths).reduce<{ [id: string]: string }>((dangers, url) => {
        const apiItem = paths[url];
        Object.keys(apiItem).forEach(method => {
          const api = apiItem[method];
          const { operationId } = api;
          if (operationId) {
            const flag = `${method.toLocaleLowerCase()} ${url}`;
            operationId2UrlMap[operationId] = flag;
            operationId2ApiMap[operationId] = url2ApiMap[flag] = api;
            const dangerousOperationId = operationId.replace(dangerousOperationIdReg, '');
            if (operationId.match(dangerousOperationIdReg)) {
              dangers[dangerousOperationId] = '';
            }
            dangerousOperationId2ApiMap[dangerousOperationId] =
              dangerousOperationId2ApiMap[dangerousOperationId] || [];
            dangerousOperationId2ApiMap[dangerousOperationId].push(api);
          }
        });
        return dangers;
      }, {})
    );
    const errors: string[] = [];
    const warnings: string[] = [];
    const suggestions: typeof operationId2UrlMap = {};
    const validateAndFix = (api: API, dangerousOperationId: string) => {
      const { operationId = '' } = api;
      // 如果未配置则出错
      if (!copySavedOperationId2UrlMap[operationId]) {
        errors.push(
          `【错误】方法 "${operationId}" & "${
            operationId2UrlMap[operationId]
          }" 与 "${dangerousOperationId}" & "${
            operationId2UrlMap[dangerousOperationId]
          }" 存在不可控风险，需锁定映射关系`
        );
        suggestions[operationId] = operationId2UrlMap[operationId];
      } else if (copySavedOperationId2UrlMap[operationId] !== operationId2UrlMap[operationId]) {
        /**
         * 尝试自动人工干涉
         *  干涉配置:
         *    get   => /api/get
         *    get_1 => /api/get-1
         *  实际返回
         *    get_1 => /api/get
         *    get   => /api/get-1
         *  干涉结果
         *    get   => /api/get
         *    get_1 => /api/get-1
         */
        const correctOpreationId = findMatchKeyByValue(
          operationId2UrlMap[operationId],
          savedOperationId2UrlMap
        );
        if (correctOpreationId) {
          warnings.push(
            `【警告】方法 "${operationId}" & "${copySavedOperationId2UrlMap[operationId]}" 与 "${
              operationId2UrlMap[operationId]
            }" 不匹配，自动纠正为 "${correctOpreationId}"`
          );
          api.operationId = correctOpreationId;
        } else {
          errors.push(
            `【错误】与 "${copySavedOperationId2UrlMap[operationId]}" 对应的方法不能为空`
          );
        }
      }
      // @cc: 风险解除
      delete copySavedOperationId2UrlMap[operationId];
    };
    dangers.reduce((errors, dangerousOperationId) => {
      const group = dangerousOperationId2ApiMap[dangerousOperationId];
      group.forEach(api => validateAndFix(api, dangerousOperationId));
      return errors;
    }, errors);
    Object.keys(copySavedOperationId2UrlMap).reduce((errors, specifiedOpreationId) => {
      // @cc: 人工干涉配置多余 swagger 配置
      const url = copySavedOperationId2UrlMap[specifiedOpreationId];
      const api = url2ApiMap[url];
      if (!api) {
        errors.push(
          `【错误】方法 "${specifiedOpreationId}" 对应的 "${url}" 似乎已被移除或者发生了变化，请更新`
        );
      } else {
        validateAndFix(api, api.operationId as string);
      }
      return errors;
    }, errors);
    rs({
      errors,
      warnings,
      suggestions
    });
  });
}

export function findMatchKeyByValue(value: any, map: { [k: string]: any }) {
  return Object.keys(map).find(key => {
    return map[key] === value;
  });
}
