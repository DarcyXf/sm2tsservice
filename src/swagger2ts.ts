import { exec } from 'child_process';
import { promisify } from 'es6-promisify';
import { SwaggerParser } from 'src';
import * as path from 'path';

const wrappedExec = <C>(url: string, cb: C) => exec(url, cb);
wrappedExec[promisify.argumentNames] = ['error', 'stdout', 'stderr'];
const asyncExec = <T>(cmd: T) =>
  promisify(wrappedExec)(cmd).then(
    (res: { error: string; stdout: string; stderr: string }) =>
      res.error ? { code: 1, message: res.stderr || res.error } : { code: 0, message: res.stdout },
    err => ({ code: 1, message: err.message })
  );

export default async function swagger2ts(
  swaggerParser: SwaggerParser
): Promise<{ code: Number; message?: string }> {
  const java = await checkJava();
  if (java.code) {
    console.error(`[ERROR]: check java failed with ${java.message}`);
    return java;
  }
  return await parseSwagger(swaggerParser);
}

async function checkJava() {
  return await asyncExec('java -version');
}
const pluginsPath = path.join(__dirname, '..', 'plugins');
const generatorPath = path.join(pluginsPath, 'swagger-codegen-cli.jar');
const defaultParseConfig = {
  '-l': 'typescript-angularjs',
  '-t': path.join(pluginsPath, 'typescript-tkit'),
  '-o': path.join(process.cwd(), 'src', 'services')
};

async function parseSwagger(swaggerParser: SwaggerParser) {
  const config = { ...defaultParseConfig, ...swaggerParser };
  return await asyncExec(
    `java -jar ${generatorPath} generate ${Object.keys(config)
      .map(opt => `${opt} ${config[opt]}`)
      .join(' ')}`
  )
    .then(() => ({ code: 0 }))
    .catch(e => ({
      code: 6,
      message: `[ERROR]: gen failed from ${config['-i']} with ${e.message}`
    }));
}
