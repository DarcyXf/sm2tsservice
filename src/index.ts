#!/usr/bin/env node
import * as fs from 'fs';
import * as path from 'path';
import swagger2ts from './swagger2ts';
import serve from './yapi/serve';

export interface Json2Service {
  url: string;
  type?: 'yapi' | 'swagger';
  swaggerParser: SwaggerParser;
}

export interface SwaggerParser {
  '-o'?: string;
  '-t'?: string;
  '-l'?: string;
  '-i': string;
}

const CD = __dirname;
const Config = 'json2service.json';
const ConfigFile = path.join(CD, '..', Config);

if (!fs.existsSync(ConfigFile)) {
  console.error(`[ERROR]: ${Config} not found in ${CD}`);
} else {
  const config: Json2Service = require(ConfigFile);
  (async () => {
    const { url, type = 'swagger', swaggerParser } = config;
    let swaggerUrl = url;
    if (type === 'yapi') {
      const yapiTMP = await serve(url);
      if ('result' in yapiTMP && yapiTMP.result && !yapiTMP.code) {
        swaggerUrl = yapiTMP.result;
      } else {
        console.error(`[ERROR]: gen failed with: ${yapiTMP.message}`);
        process.exit();
      }
    }
    const res = await swagger2ts({ ...swaggerParser, '-i': swaggerUrl });
    if (res.code) {
      console.error(`[ERROR]: gen failed with: ${res.message}`);
    } else {
      console.log(`[INFO]: gen success with: ${url}`);
    }
  })();
}
