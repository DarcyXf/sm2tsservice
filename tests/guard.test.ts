/* eslint-disable @typescript-eslint/camelcase */
import { operationIdGuard } from 'src/guard';
import { SwaggerJson } from 'src/consts';

const url = '/api/v1/persons';
const url1 = '/api/v1/interviewer/persons';
function getSwagger(): SwaggerJson {
  return {
    basePath: '/',
    paths: {
      [url1]: {
        get: {
          operationId: 'personsUsingGET',
          parameters: [],
          responses: {
            '200': {
              description: 'OK'
            }
          },
          deprecated: false
        }
      },
      [url]: {
        get: {
          operationId: 'personsUsingGET_1',
          parameters: [],
          responses: {
            '200': {
              description: 'OK'
            }
          },
          deprecated: false
        }
      }
    }
  };
}

const optionIdMethodUrlMap = {
  personsUsingGET: 'get /api/v1/persons',
  personsUsingGET_1: 'get /api/v1/interviewer/persons'
};

const optionIdMethodUrlMap2 = {
  personsUsingGET: 'get /api/v1/interviewer/persons',
  personsUsingGET_1: 'get /api/v1/persons'
};

describe('validate/utils', () => {
  it('operationIdGuard should work ok', async () => {
    let swagger = getSwagger();
    let res = await operationIdGuard(swagger, {});
    expect(res).toMatchSnapshot('error');
    expect(swagger).toMatchSnapshot('error');

    swagger = getSwagger();
    res = await operationIdGuard(swagger, optionIdMethodUrlMap);
    expect(res).toMatchSnapshot('autofix');
    expect(swagger).toMatchSnapshot('autofix');

    swagger = getSwagger();
    swagger.paths = {};
    res = await operationIdGuard(swagger, optionIdMethodUrlMap);
    expect(res).toMatchSnapshot('remove error');
    expect(swagger).toMatchSnapshot('remove error');

    swagger = getSwagger();
    swagger.paths = {
      [url1]: swagger.paths[url1]
    };
    res = await operationIdGuard(swagger, optionIdMethodUrlMap);
    expect(res).toMatchSnapshot('partial remove error');
    expect(swagger).toMatchSnapshot('partial remove error');

    swagger = getSwagger();
    swagger.paths = {
      [url1]: swagger.paths[url1]
    };
    res = await operationIdGuard(swagger, optionIdMethodUrlMap2);
    expect(res).toMatchSnapshot('partial remove ok');
    expect(swagger).toMatchSnapshot('partial remove ok');
  });
});
