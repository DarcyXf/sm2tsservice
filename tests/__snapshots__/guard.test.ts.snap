// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validate/utils operationIdGuard should work ok: 1 error 1`] = `
Object {
  "errors": Array [
    "【错误】方法 \\"personsUsingGET\\" & \\"get /api/v1/interviewer/persons\\" 存在不可控风险，需锁定映射关系",
    "【错误】方法 \\"personsUsingGET_1\\" & \\"get /api/v1/persons\\" 存在不可控风险，需锁定映射关系",
  ],
  "suggestions": Object {
    "get /api/v1/interviewer/persons": "personsUsingGET",
    "get /api/v1/persons": "personsUsingGET_1",
  },
  "warnings": Array [],
}
`;

exports[`validate/utils operationIdGuard should work ok: 1 error 2`] = `
Object {
  "basePath": "/",
  "paths": Object {
    "/api/v1/interviewer/persons": Object {
      "get": Object {
        "deprecated": false,
        "operationId": "personsUsingGET",
        "parameters": Array [],
        "privateMethodUrl": "get /api/v1/interviewer/persons",
        "responses": Object {
          "200": Object {
            "description": "OK",
          },
        },
      },
    },
    "/api/v1/persons": Object {
      "get": Object {
        "deprecated": false,
        "operationId": "personsUsingGET_1",
        "parameters": Array [],
        "privateMethodUrl": "get /api/v1/persons",
        "responses": Object {
          "200": Object {
            "description": "OK",
          },
        },
      },
    },
  },
}
`;

exports[`validate/utils operationIdGuard should work ok: 2 autofix 1`] = `
Object {
  "errors": Array [],
  "suggestions": Object {},
  "warnings": Array [
    "【警告】方法 \\"personsUsingGET\\" 与 \\"get /api/v1/interviewer/persons\\" 不匹配，自动纠正为 \\"personsUsingGET_1\\"",
    "【警告】方法 \\"personsUsingGET_1\\" 与 \\"get /api/v1/persons\\" 不匹配，自动纠正为 \\"personsUsingGET\\"",
  ],
}
`;

exports[`validate/utils operationIdGuard should work ok: 2 autofix 2`] = `
Object {
  "basePath": "/",
  "paths": Object {
    "/api/v1/interviewer/persons": Object {
      "get": Object {
        "deprecated": false,
        "operationId": "personsUsingGET_1",
        "parameters": Array [],
        "responses": Object {
          "200": Object {
            "description": "OK",
          },
        },
      },
    },
    "/api/v1/persons": Object {
      "get": Object {
        "deprecated": false,
        "operationId": "personsUsingGET",
        "parameters": Array [],
        "responses": Object {
          "200": Object {
            "description": "OK",
          },
        },
      },
    },
  },
}
`;

exports[`validate/utils operationIdGuard should work ok: 3 remove error 1`] = `
Object {
  "errors": Array [
    "【错误】方法 \\"personsUsingGET\\" 对应的 \\"get /api/v1/persons\\" 似乎已被移除或者发生了变化，请更新",
    "【错误】方法 \\"personsUsingGET_1\\" 对应的 \\"get /api/v1/interviewer/persons\\" 似乎已被移除或者发生了变化，请更新",
  ],
  "suggestions": Object {},
  "warnings": Array [],
}
`;

exports[`validate/utils operationIdGuard should work ok: 3 remove error 2`] = `
Object {
  "basePath": "/",
  "paths": Object {},
}
`;

exports[`validate/utils operationIdGuard should work ok: 4 partial remove error 1`] = `
Object {
  "errors": Array [
    "【错误】方法 \\"personsUsingGET\\" 对应的 \\"get /api/v1/persons\\" 似乎已被移除或者发生了变化，请更新",
  ],
  "suggestions": Object {},
  "warnings": Array [
    "【警告】方法 \\"personsUsingGET\\" 与 \\"get /api/v1/interviewer/persons\\" 不匹配，自动纠正为 \\"personsUsingGET_1\\"",
  ],
}
`;

exports[`validate/utils operationIdGuard should work ok: 4 partial remove error 2`] = `
Object {
  "basePath": "/",
  "paths": Object {
    "/api/v1/interviewer/persons": Object {
      "get": Object {
        "deprecated": false,
        "operationId": "personsUsingGET_1",
        "parameters": Array [],
        "responses": Object {
          "200": Object {
            "description": "OK",
          },
        },
      },
    },
  },
}
`;

exports[`validate/utils operationIdGuard should work ok: 5 partial remove ok 1`] = `
Object {
  "errors": Array [
    "【错误】方法 \\"personsUsingGET_1\\" 对应的 \\"get /api/v1/persons\\" 似乎已被移除或者发生了变化，请更新",
  ],
  "suggestions": Object {},
  "warnings": Array [],
}
`;

exports[`validate/utils operationIdGuard should work ok: 5 partial remove ok 2`] = `
Object {
  "basePath": "/",
  "paths": Object {
    "/api/v1/interviewer/persons": Object {
      "get": Object {
        "deprecated": false,
        "operationId": "personsUsingGET",
        "parameters": Array [],
        "privateMethodUrl": "get /api/v1/interviewer/persons",
        "responses": Object {
          "200": Object {
            "description": "OK",
          },
        },
      },
    },
  },
}
`;

exports[`validate/utils strictModeGuard should work ok: strictModeGuard 1`] = `
Object {
  "errors": Array [
    "【错误】 tags \\"{
  \\"name\\": \\"错误\\",
  \\"description\\": \\"这个tags不合格\\"
}\\" 命名包含非英文字符，请联系接口方修改",
    "【错误】 tags \\"{
  \\"name\\": \\"xxx-controller\\",
  \\"description\\": \\"这个tags不合格\\"
}\\" 命名不符合规范或由于 @Api 未添加 tags，存在不可控风险，请联系接口方修改",
  ],
  "warnings": Array [],
}
`;
